### PR TITLE
Issue #1994: Hide the registration username field if users can only login with email

### PR DESCRIPTION
**Pull request by docwilmot**
Fixes https://github.com/backdrop/backdrop-issues/issues/1994

### CODE REVIEW

**Comment by docwilmot** [comment]
Sandbox site is inaccessible as someone saved the option to not use usernames and we dont get an email address on the sandbox! Reopening.

**Comment by docwilmot** [comment]
The email address for this new sandbox is:

testing@1763.backdrop.backdrop.qa.backdropcms.org

If you save the option to *not* use usernames, you will need the email to sign back in.

File: core/modules/user/user.module [hunk 1]
```diff
@@ -689,6 +689,9 @@ function user_account_form(&$form, &$form_state) {
     '#type'   => 'container',
     '#weight' => -10,
   );
+
+  $hide_username = config_get('system.core', 'hide_username_on_registration');
```
Comment (docwilmot) [inline core/modules/user/user.module:693]: Need to add a default to the `config.json` file.

File: core/modules/user/user.admin.inc [hunk 2]
```diff
@@ -137,6 +137,35 @@ function user_admin_settings($form, &$form_state) {
...
+  );
+  $form['user_login']['hide_username_wrapper']['username_generation_method'] = array(
+    '#type' => 'radios',
+    '#title' => t('Generate usernames'),
+    '#options' => array(
+      0 => t('Use the email address as username'),
+      1 => t('Generate a username based on the email address'),
+      2 => t('Generate a random username'),
+    ),
+    '#default_value' => $config->get('username_generation_method'),
```
Comment (docwilmot) [inline core/modules/user/user.admin.inc:162]: Need to add a default to the `config.json` file.

File: core/modules/user/user.module [hunk 3]
```diff
@@ -872,6 +875,24 @@ function user_account_form_validate($form, &$form_state) {
   // Validate new or changing username.
   if (isset($form_state['values']['name'])) {
     $name = $form_state['values']['name'];
+    $hide_username_on_registration = config_get('system.core', 'hide_username_on_registration');
+    // If login is by email only, admins have the option to hide the username
+    // field. If it's hidden, set the username as the email during registration.
+    if ($name == '' && $hide_username_on_registration && empty($account->uid)) {
+      $username_generation_method = config_get('system.core', 'username_generation_method');
+      switch ($username_generation_method) {
+        case 0:
```
Comment (docwilmot) [inline core/modules/user/user.module:884]: Should use strings rather than integer to make readable in `config.json` file.

**Comment by docwilmot** [comment]
This also needs tests. 
P.s.: yes, I am code reviewing my own PR here. Just reminding myself as no access to GitHub here.

File: core/modules/user/user.module (hunk 1 above)
Comment (jenlampton) [reply core/modules/user/user.module:693]: added

**Comment by jenlampton** [comment]
Replaced by https://github.com/backdrop/backdrop/pull/1856

**Comment by jenlampton** [comment]
closing in favor of an alternate solution. See backdrop/backdrop-issues#2669

**Pull request closed without merging**